git/revlist.go

```diff
@@ -117,3 +120,207 @@ func revListCallback(c *Client, opt RevListOptions, commits []CommitID, excludeL
 	}
 	return nil
 }
+
+// Internal pre-computed record of which symbolic references
+//  are pointing to a particular ref name (without the refs/ prefix)
+var symRefsForRefName map[string]string
+
+// Internal pre-computed record of which ref names are pointing to
+//  to a particular commit. Values are in the form of "HEAD -> master, branch2,
+//  tag: tag1"
+var refNamesForCommit map[string]string
```

Comment:
Are you sure that these are required, or are you just assuming they are? I know I had to rip out some caches (mostly with delta calculation) a while ago because while they helped with smaller repos, they ate up all the memory and crashed on bigger repos.

If they are required, I think that:
1. They should use better types than strings
2. They should be on the Client object instead of globals for thread safety.

Reply:
Agreed that this could be premature optimization without any benchmarks. I felt that reading the ref and head files over and over for each commit in history might be bad performance, but perhaps I am underestimating OS caches.

Fair enough, I can remove these for now until we notice a performance problem. If they are needed, the caches could be managed and typed better.

Reply:
You're probably right, and caching every reference is cheaper than caching every commit (which is where we had problems before).. but I still think it should be on Client so that multiple client's can use the library concurrently without bumping into each other's cache.